### PR TITLE
Implementation of the option report_changes for Windows registries

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/5865|

## Description
Branch work for the implementation of the report_changes option in the Windows registry improvement.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
- [x] Source installation

- Memory tests for Linux
  - [x] Scan-build report
- Memory tests for Windows
  - [x] Scan-build report

### CODE REVIEW

File: src/syscheckd/fim_diff_changes.c [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+        written += snprintf(adapted_output + written, OS_MAXSTR - written, "%s%s%s", line_mode == 1 ? "< " : "> ", line, next_line ? "\n" : "");
+
+next_it:
+        line = next_line;
+    }
+
+    return adapted_output;
+}
+
+int fim_diff_process_delete_file(char *filename){
```
Comment (Molter73) [inline src/syscheckd/fim_diff_changes.c:1017]:
```suggestion
int fim_diff_process_delete_file(const char *filename){
```

File: src/syscheckd/fim_diff_changes.c [hunk 2]
```diff
@@ -0,0 +1,1096 @@
...
+    }
+
+    if(fim_diff_delete_compress_folder(full_path) == -1){
+        os_free(full_path);
+        return -1;
+    }
+
+    os_free(full_path);
+    return 0;
+}
```
Comment (Molter73) [inline src/syscheckd/fim_diff_changes.c:1070]: Consider the following alternate implementation of this function:
``` c
int fim_diff_process_delete_registry(const char *key_name, int arch) {
    char full_path[PATH_MAX];
    os_sha1 encoded_key;

    OS_SHA1_Str(key_name, strlen(key_name), encoded_key);

    if (arch) {
        snprintf(full_path, PATH_MAX, "%s/registry/[x64] %s", DIFF_DIR_PATH, encoded_key);
    } else {
        snprintf(full_path, PATH_MAX, "%s/registry/[x32] %s", DIFF_DIR_PATH, encoded_key);
    }

    if (fim_diff_delete_compress_folder(full_path) == -1) {
        return -1;
    }

    return 0;
}
```

File: src/syscheckd/fim_diff_changes.c [hunk 3]
```diff
@@ -0,0 +1,1096 @@
...
```
Comment (Molter73) [inline src/syscheckd/fim_diff_changes.c:1094]: Consider the following alternate implementation of this function:
``` c
int fim_diff_process_delete_value(const char *key_name, const char *value_name, int arch) {
    char full_path[PATH_MAX];
    os_sha1 encoded_key;
    os_sha1 encoded_value;

    OS_SHA1_Str(key_name, strlen(key_name), encoded_key);
    OS_SHA1_Str(value_name, strlen(value_name), encoded_value);

    if (arch) {
        snprintf(full_path, PATH_MAX, "%s/registry/[x64] %s/%s", DIFF_DIR_PATH, encoded_key, encoded_value);
    } else {
        snprintf(full_path, PATH_MAX, "%s/registry/[x32] %s/%s", DIFF_DIR_PATH, encoded_key, encoded_value);
    }

    if (fim_diff_delete_compress_folder(full_path) == -1) {
        return -1;
    }

    return 0;
}
```

File: src/syscheckd/syscheck.h [hunk 4]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
  * @param value_name Name of the value that has generated the alert
  * @param value_data Content of the value to be checked
  * @param data_type The type of value we are checking
- * @return String with the diff to add to the alert
+ * @param registry Config of the registry key
+ * @return String with the changes to add to the alert
  */
 
-char * fim_registry_value_diff(char *key_name, char *value_name, LPBYTE value_data, DWORD data_type);
+char * fim_registry_value_diff(char *key_name, char *value_name, char *value_data, DWORD data_type, registry *configuration);
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:601]:
```suggestion
char *fim_registry_value_diff(const char *key_name,
                              const char *value_name,
                              const char *value_data,
                              DWORD data_type,
                              const registry *configuration);
```

File: src/syscheckd/syscheck.h [hunk 5]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
...
+
+char * fim_file_diff(char *filename);
+
+/**
+ * @brief Deletes the filename diff folder and modify diff_folder_size if disk_quota enabled
+ *
+ * @param filename Path of the file that has been deleted
+ * @return 0 if success, -1 on error
+ */
+int fim_diff_process_delete_file(char *filename);
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:619]:
```suggestion
int fim_diff_process_delete_file(const char *filename);
```

File: src/syscheckd/syscheck.h [hunk 6]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
...
+
+/**
+ * @brief Deletes the registry diff folder and modify diff_folder_size if disk_quota enabled
+ *
+ * @param key_name Path of the registry that has been deleted
+ * @param arch Arch type of the registry
+ * @return 0 if success, -1 on error
+ */
+int fim_diff_process_delete_registry(char *key_name, int arch);
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:628]:
```suggestion
int fim_diff_process_delete_registry(const char *key_name, int arch);
```

File: src/syscheckd/syscheck.h [hunk 7]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
...
  *
- * @param filename The name of the file to check
- * @return 1 if the file has been configured with the ``nodiff`` option, 0 if not
+ * @param key_name Path of the registry that contain value deleted
+ * @param value_name Path of the value that has been deleted
+ * @param arch Arch type of the registry
+ * @return 0 if success, -1 on error
  */
-int is_nodiff(const char *filename);
+int fim_diff_process_delete_value(char *key_name, char *value_name, int arch);
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:638]:
```suggestion
int fim_diff_process_delete_value(const char *key_name, const char *value_name, int arch);
```

File: src/syscheckd/syscheck.h [hunk 8]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
...
+ * @brief Deletes the registry diff folder and modify diff_folder_size if disk_quota enabled
+ *
+ * @param key_name Path of the registry that has been deleted
+ * @param arch Arch type of the registry
+ * @return 0 if success, -1 on error
+ */
+int fim_diff_process_delete_registry(char *key_name, int arch);
+
+/**
+ * @brief Deletes the value diff folder and modify diff_folder_size if disk_quota enabled
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:631]:
```suggestion
 * @brief Deletes the value diff folder and modifies diff_folder_size if disk_quota enabled
```

File: src/syscheckd/syscheck.h [hunk 9]
```diff
@@ -646,19 +594,48 @@ long unsigned int WINAPI state_checker(__attribute__((unused)) void *_void);
...
  *
- * @param filename The name of the file to check
- * @return 1 if the file has been configured with the ``nodiff`` option, 0 if not
+ * @param key_name Path of the registry that contain value deleted
```
Comment (Molter73) [inline src/syscheckd/syscheck.h:633]:
```suggestion
 * @param key_name Path of the registry that contains the deleted value
```

File: src/config/syscheck-config.c [hunk 10]
```diff
@@ -375,6 +375,61 @@ int dump_registry_ignore_regex(syscheck_config *syscheck, char *regex, int arch)
     return 1;
 }
 
+void dump_registry_nodiff(syscheck_config *syscheck, char *entry, int arch) {
```
Comment (Molter73) [inline src/config/syscheck-config.c:378]:
```suggestion
void dump_registry_nodiff(syscheck_config *syscheck, const char *entry, int arch) {
```

File: src/config/syscheck-config.c [hunk 11]
```diff
@@ -375,6 +375,61 @@ int dump_registry_ignore_regex(syscheck_config *syscheck, char *regex, int arch)
...
+        os_calloc(2, sizeof(registry), syscheck->registry_nodiff);
+        syscheck->registry_nodiff[0].entry = NULL;
+        syscheck->registry_nodiff[1].entry = NULL;
+    }
+
+    os_strdup(entry, syscheck->registry_nodiff[ign_size].entry);
+    syscheck->registry_nodiff[ign_size].arch = arch;
+}
+
+int dump_registry_nodiff_regex(syscheck_config *syscheck, char *regex, int arch) {
```
Comment (Molter73) [inline src/config/syscheck-config.c:402]:
```suggestion
int dump_registry_nodiff_regex(syscheck_config *syscheck, const char *regex, int arch) {
```

File: src/config/syscheck-config.c [hunk 12]
```diff
@@ -375,6 +375,61 @@ int dump_registry_ignore_regex(syscheck_config *syscheck, char *regex, int arch)
...
+        os_calloc(2, sizeof(registry_regex), syscheck->registry_nodiff_regex);
+        syscheck->registry_nodiff_regex[0].regex = NULL;
+        syscheck->registry_nodiff_regex[1].regex = NULL;
+    } else {
+        while (syscheck->registry_nodiff_regex[ign_size].regex != NULL) {
+            ign_size++;
+        }
+
+        os_realloc(syscheck->registry_nodiff_regex, sizeof(registry_regex) * (ign_size + 2),
+                syscheck->registry_nodiff_regex);
```
Comment (Molter73) [inline src/config/syscheck-config.c:416]:
```suggestion
                   syscheck->registry_nodiff_regex);
```

File: src/config/syscheck-config.c [hunk 13]
```diff
@@ -375,6 +375,61 @@ int dump_registry_ignore_regex(syscheck_config *syscheck, char *regex, int arch)
...
+        syscheck->registry_nodiff_regex[ign_size + 1].regex = NULL;
+    }
+
+    os_calloc(1, sizeof(OSMatch),
+            syscheck->registry_nodiff_regex[ign_size].regex);
```
Comment (Molter73) [inline src/config/syscheck-config.c:421]:
```suggestion
    os_calloc(1, sizeof(OSMatch), syscheck->registry_nodiff_regex[ign_size].regex);
```

**Review by Molter73 — CHANGES_REQUESTED** [review]

File: src/syscheckd/fim_diff_changes.c [hunk 14]
```diff
@@ -0,0 +1,1096 @@
...
+    os_free(diff->compress_tmp_file);
+    os_free(diff->diff_file);
+
+    os_free(diff);
+
+    return;
+}
+
+bool fim_diff_check_limits(diff_data *diff, char *filename) {
+    diff->file_size = (float)FileSize(diff->file_origin) / 1024;
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply src/syscheckd/fim_diff_changes.c:608]: This variable is used in the following "if" block as well. It is always necessary to initialize

File: src/config/syscheck-config.c [hunk 15]
```diff
@@ -1458,6 +1585,15 @@ int Read_Syscheck(const OS_XML *xml, XML_NODE node, void *configp, __attribute__
...
                             return OS_INVALID;
                         }
+                    } else if (strcmp(node[i]->attributes[j], xml_report_changes) == 0) {
+                        if (strcmp(node[i]->values[j], "yes") == 0) {
+                            opts |= CHECK_SEECHANGES;
+                        } else if (strcmp(node[i]->values[j], "no") == 0) {
+                            opts &= ~ CHECK_SEECHANGES;
+                        } else {
+                            mwarn(FIM_INVALID_REG_OPTION_SKIP, node[i]->values[j], node[i]->attributes[j], node[i]->content);
+                            return OS_INVALID;
```
Comment (antoniomanuelfr) [inline src/config/syscheck-config.c:1595]: The memory pointed by tag is not freed before returning `OS_INVALID`. 
This is the only bug reported by scan-build
```suggestion
                            mwarn(FIM_INVALID_REG_OPTION_SKIP, node[i]->values[j], node[i]->attributes[j], node[i]->content);
                            os_free(tag)
                            return OS_INVALID;
```

File: src/shared/syscheck_op.c [hunk 16]
```diff
@@ -113,7 +82,7 @@ int remove_empty_folders(const char *path) {
                 if (rmdir_ex(parent) != 0) {
                     mwarn("Empty directory '%s' couldn't be deleted. ('%s')",
                         parent, strerror(errno));
-                    retval = 1;
+                    retval = -1;
```
Comment (antoniomanuelfr) [inline src/shared/syscheck_op.c:85]: Fix indentation.

File: src/config/syscheck-config.c [hunk 17]
```diff
@@ -1458,6 +1585,15 @@ int Read_Syscheck(const OS_XML *xml, XML_NODE node, void *configp, __attribute__
                             os_free(tag);
                             return OS_INVALID;
                         }
+                    } else if (strcmp(node[i]->attributes[j], xml_report_changes) == 0) {
+                        if (strcmp(node[i]->values[j], "yes") == 0) {
+                            opts |= CHECK_SEECHANGES;
+                        } else if (strcmp(node[i]->values[j], "no") == 0) {
+                            opts &= ~ CHECK_SEECHANGES;
+                        } else {
+                            mwarn(FIM_INVALID_REG_OPTION_SKIP, node[i]->values[j], node[i]->attributes[j], node[i]->content);
```
Comment (antoniomanuelfr) [inline src/config/syscheck-config.c:1594]: The memory pointed by `tag` is not freed.
This is the only bug reported by scan-build.
```suggestion
                            mwarn(FIM_INVALID_REG_OPTION_SKIP, node[i]->values[j], node[i]->attributes[j], node[i]->content);
                            os_free(tag);
```

**Review by antoniomanuelfr — CHANGES_REQUESTED** [review]

**Review by antoniomanuelfr — APPROVED** [review]
LGTM!

**Review by Molter73 — APPROVED** [review]
LGTM!

**Pull request merged**